Color histogram with floats

zoccoler opened the pull request:
This adds the possibility to assign an array with floats (like representing some feature) as an overlay to the histogram and to the scatter plot.
- [x] update documentation
- [x] check if more tests are needed
- [x] fix log scale problem for histogram

zoccoler commented:
Hi @jo-mueller ,

I am adding here the possibility to display the scatter colors or an histogram overlay based o another features column (not just a categorical array with integers, but rather any float array with a continuous colormap - I renamed `categorical_colormap` to `overlay_colormap`).

You can get a feeling of the changes by running the Artists notebook again, which I updated.

Of course a histogram overlay completely hides the original histogram, but it may be useful if the user hide and show it quickly or change its opacity for example. 

When you get some time, could you please give it a try to check if we can merge this?

jo-mueller commented:
Hi @zoccoler ,

sorry for the late reply. I am a bit undecided on this implementation: On the one hand I like the simplicity of this: Simply change the colormap to something continuous, pass a float to the color indices and have your plot colored. On the other hand, it loads the same property (color) with potentially multiple meanings. 

In the context of the clusters plotter, it would be hard to use colored scatter points in conjunction with the clustering. As an alternative suggestion: Would you think that it could make sense to forward the coloring done with `color_indeces` to the marker edge color? The coloring of the points according to a given feature could then still be done with the facecolor and the two colorings wouldn't superseed each other.

Let me know what you think!

zoccoler commented:
Hi @jo-mueller ,

I think we both agree that displaying a feature with different colors is a good enhancement (it also fits the goal of [this PR](https://github.com/BiAPoL/napari-clusters-plotter/pull/257).

Using the edge colors for categories can indeed be a solution, we just have to test if it doesn't look confusing. This wouldn't work for the histogram though. 
I was hoping to have some hide/show overlay option, similarly to one showing a labels layer on top of an image layer in napari. That would work for both the histogram and the scatter artists.

I will try to generate some examples for these ideas here, then we can decide.

zoccoler commented:
Hi @jo-mueller ,

I have added a button to hide/show the colors overlay (for histogram or scatter).

I also added a signal emitted when the `color_indices` are updated after a selection. That should allow us to differentiate when a user updates the coloring via the "hue" combobox in clusters-plotter or after a manual selection.
If done via manual selection, the default colormap of the `overlay_colormap` is restored and `color_indices` are reset to avoid wrong leftovers from previous coloring.

What do you think? I didn't try the edge_color approach yet.

jo-mueller commented on src/biaplotter/artists.py:
```diff
@@ -236,26 +247,104 @@ def color_indices(self, indices: np.ndarray):
+    def overlay_colormap(self, value: Colormap):
+        """Sets the overlay colormap for the scatter plot."""
+        self._overlay_colormap = BiaColormap(value)
+        self.color_indices = self._color_indices
```
I think this ends up being problematic for me in some points. Why does changing the colormap require setting the `color_indices`? In the scope of the napari clusters plotter, this triggers an infinite loop.

Cascade:
1. color_indeces are changed (e.g., when a hue value is selected)
2. `color_indeces_changed` is emitted
3. layer recoloring is triggered
4. layer coloring inspects the hue value to color by and changes the colormap of the plot accordingly, if necessary
5. `color_indeces` is overwritten here
6. `color_indeces_changed` is emitted --> infinte loop

jo-mueller replied:
```suggestion

```

jo-mueller commented on src/biaplotter/artists.py:
```diff
@@ -236,26 +247,104 @@ def color_indices(self, indices: np.ndarray):
+            # Create a ScalarMappable instance using chosen colormap and normalization
+            sm = ScalarMappable(norm=norm, cmap=self.overlay_colormap.cmap)
+            # Convert normalized data to RGBA
+            rgba_colors = sm.to_rgba(indices)
```
This is problematic, I think. By applying a color normalization here, there is no longer a clear mapping between `color_indices` and the actually drawn color in the plot.

Usually, what should work is 
```python
color = active_artist.overlay_colormap(color_indices)
```
and the colors I get should be exactly the same as in the plot. However, the colors in the plot pass through an additional normalization step. To recover the colors from the plot (like needed for the clusters plotter), I need to do this:

```python
        color_indices = active_artist.color_indices
        norm = active_artist._get_normalization(color_indices)
        colors = active_artist._get_rgba_colors(color_indices, norm)
```

Maybe there could be an additional convenience method to wrap the normalization and `to_rgba` step? Or the normalization step could go into the `__call__` method of the new `BiaColormap`?

zoccoler commented:
> Hi @zoccoler , besides the above points, another problem for me is that with the `color_indeces` now also carrying continuous values, I cannot distinguish anymore between these two cases:
> 
> * A cluster was drawn on the plotter canvas and the cluster indeces are emitted
> * Some continuous values were passed to the plotter and the values are subsequently emitted
> 
> In the scope of the clusters-plotter, this distinction is important though. Consider this workflow:
> 
> * I plot some continuous feature in the clusters-plotter
> * When I start drawing and finish, the clusters should reset so that the `MANUAL_CLUSTER_ID` is plotted again.
> * In this case, the `hue` selector in the widget should jump to `MANUAL_CLUSTER_ID` as well.
> * I have no way of knowing to which value to set the `hue` selector from the `cluster_indeces`.

That is why I had placed [this](https://github.com/BiAPoL/biaplotter/blob/15974f7fb54b0209d6917dbbbe967b31ced5334e/src/biaplotter/selectors.py#L347) 😅 Does it help to solve that problem?

jo-mueller commented:
@zoccoler , just tested it, the new signal works on my end for clusters-plotter purposes. Can you have a look at PR #42 so that this PR can go through?

zoccoler commented:
Hi @jo-mueller ,

This is now merged. I will make the new release early next week because I have to prepare a fix release for napari-phasors due to the breaking changes introduced here.

Please refer to the main branch for now while working in the clusters-plotter.
